File: java/palindrome_linked_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode one=head, two=head, curr=null, prev=null, next=null;
        if (head==null) return true;
        while (two!=null && two.next!=null) {
            prev=one;
            one=one.next;
            two=two.next.next;
        }
        
        if (prev!=null) prev.next=null;
        
        prev=null;
        curr=one;
        while (curr!=null) {
            next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }
        
        ListNode l1=prev, l2=head;
        while (l2!=null) {
            if (l1.val!=l2.val) return false;
            l1=l1.next;
            l2=l2.next;
        }
        return true;
    }
}

File: cpp/palindrome_linked_list.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <vector>
#include <queue>
#include <stack>
#include <list>
#include <set>
#include <unordered_set>
#include <map>
#include <algorithm>
#include <limits.h>
#include <math.h>
#include <iostream>
#include <sstream>

  class Solution {
  public:
      bool isPalindrome(ListNode* head) {
    	  ListNode *one=head, *two=head;
    	  ListNode *prev=NULL;
    	  while (two!=NULL && two->next!=NULL) {
    		  one=one->next;
    		  two=two->next->next;
    	  }

    	  // reverse from one
    	  while (one!=NULL) {
    		  ListNode *next=one->next;
    		  one->next=prev;
    		  prev=one;
    		  one=next;
    	  }

    	  ListNode *front=head, *back=prev;
    	  while (back!=NULL) {
    		  if (back->val!=front->val) return false;
    		  front=front->next;
    		  back=back->next;
    	  }
    	  return true;
      }
  };

int main()
{
	return 0;
}

